src/native/pprof.c

```diff
@@ -273,21 +269,67 @@ static uint64_t _pprof_locNew(DProf* dp, uint64_t id_mapping, uint64_t addr, uin
 
 }
 
-uint64_t pprof_locAdd(DProf* dp, uint64_t id_mapping, uint64_t addr, uint64_t* functions, int64_t* lines, size_t n_functions) {
+static bool isEqualLine(const PPLine *line, uint64_t function_id, int64_t lineno) {
+  uint64_t fid_a = line->function_id, fid_b = function_id;
+  int64_t line_a = line->line, line_b = lineno;
+  return !((fid_a ^ fid_b) | (line_a ^ line_b));
+}
+
+static bool isEqualLocation(const PPLocation *location, uint64_t mapping_id,
+                            uint64_t addr, const uint64_t *functions,
+                            const int64_t *lines, size_t n_functions) {
+  if (location->mapping_id != mapping_id) return false;
+  if (location->address != addr) return false;
+  if (location->n_line != n_functions) return false;
+
+  for (size_t i = 0; i < n_functions; ++i) {
+    if (!isEqualLine(location->line[i], functions[i], lines[i])) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+static PPLocation *findLocation(DProf* dp, uint64_t id_mapping, uint64_t addr,
+                                const uint64_t *functions, const int64_t *lines,
+                                size_t n_functions) {
   PPProfile* pprof = &dp->pprof;
-  for(size_t i=0; i < pprof->n_location; i++) {
-    if(isEqualLocation(id_mapping, addr, pprof->location[i]))
-      return i+1;
+  for (size_t i = 0; i < pprof->n_location; ++i) {
+    PPLocation *location = pprof->location[i];
+    if (isEqualLocation(location, id_mapping, addr, functions, lines, n_functions)) {
+      return location;
+    }
   }
-  return _pprof_locNew(dp, id_mapping, addr, functions, lines, n_functions) + 1;
+  return NULL;
 }
 
-char isEqualSample(uint64_t* loc, size_t nloc, PPSample* B) {
-  if (nloc != B->n_location_id) return 0;
-  for (size_t i = 0; i <= nloc; i++) {
-    if (loc[i] != B->location_id[i]) return 0;
+uint64_t pprof_locAdd(DProf* dp, uint64_t id_mapping, uint64_t addr,
+                      const uint64_t *functions, const int64_t *lines,
+                      size_t n_functions) {
+  PPLocation *location = findLocation(dp, id_mapping, addr, functions, lines, n_functions);
+  return (location)
```

Comment:
Is splitting the conditional operator along the front of the line LLVM style?  Am I just a goofy old bat for immediately parsing 311 as `return location;`?

If there aren't any strong stylistic reservations, then splitting operators at the end of lines ensures the reader knows that they need to keep parsing to get the full statement (OK, yeah, formally we all need to scan for semicolons, but presence-of-token is easier for humans than lack-of-token in general)

```
  return (location) ?
             location -> id :
             ...
```


Again, no particularly strong feelings from me, just noting an ever-so-slight consideration for an alternative.

Reply:
I wrote it that style, actually. I haven't seen it with the operators at the end-of-line before, I don't think. Here's how `clang-format` does it:

```
  return (location) ? location->id
                    : _pprof_locNew(dp, id_mapping, addr, functions, lines,
                                    n_functions) +
                          1;
```

Which is... bad. Its gratuitous indenting means it line wrapped.  I probably should tweak the expression a bit so it gives a better formatting, or see what knobs there are to control ternaries.